---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r 'setup', include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
library('BiocStyle')
library('spatialLIBD')
## Bib setup
library('knitcitations')
## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')
# Note links won't show for now due to the following issue
# https://github.com/cboettig/knitcitations/issues/63
## Write bibliography information
bib <- c(
    R = citation(),
    AnnotationHub = citation('AnnotationHub'),
    BiocStyle = citation('BiocStyle'),
    cowplot = citation('cowplot'),
    DT = citation('DT'),
    ExperimentHub = citation('ExperimentHub'),
    fields = citation('fields'),
    ggplot2 = citation('ggplot2'),
    golem = citation('golem'),
    IRanges = citation('IRanges'),
    knitcitations = citation('knitcitations'),
    knitr = citation('knitr')[3],
    plotly = citation('plotly'),
    Polychrome = citation('Polychrome'),
    RColorBrewer = citation('RColorBrewer'),
    rmarkdown = citation('rmarkdown')[1],
    S4Vectors = citation('S4Vectors'),
    scater = citation('scater'),
    sessioninfo = citation('sessioninfo'),
    SingleCellExperiment = citation('SingleCellExperiment'),
    shiny = citation('shiny'),
    SummarizedExperiment = citation('SummarizedExperiment'),
    testthat = citation('testthat'),
    viridisLite = citation('viridisLite'),
    spatialLIBD = citation('spatialLIBD')[1],
    spatialLIBDpaper = citation('spatialLIBD')[2]
)
```

# Spatial DLPFC

This project has the code for the project using 10X Genomics Visium for generating spatial transcriptomics data from the dorsolateral prefrontal cortex (DLPFC) from human brain samples. This work is being done by the groups at the Lieber Institute for Brain Development led by Keri Martinowich, Kristen Maynard and Leonardo Collado-Torres as well as Stephanie Hicks from JHBSPH's Biostatistics Department.

# Internal

_Note: the idea is to use the same structure as in the [`locus-c`](https://github.com/lmweber/locus-c) project._

* JHPCE location: `/dcs04/lieber/lcolladotor/spatialDLPFC_LIBD4035/spatialDLPFC`.
* Slack channel: `libd_dlpfc_spatial`.

## Files

- `code`: scripts for running `SpaceRanger` and doing the analyses.
- `plots`: plots generated by RMarkdown or R analysis scripts in `.pdf` or `.png` format
- `processed-data`
  * `images_spatialLIBD`: images used for running `SpaceRanger`
  * `NextSeq`: `SpaceRanger` output files
  * `rdata`: R objects
- `raw-data`
  * `FASTQ`: FASTQ files from `NextSeq` runs. 
  * `Images`: raw images from the scanner in `.tif` format and around 3 GB per sample.
  * `images_raw_align_json`
  * `sample_info`: spreadsheet with information about samples (sample ID, sample name, slide serial number, capture area ID)
  
This project is organized along the guidelines at https://lcolladotor.github.io/bioc_team_ds/organizing-your-work.html#.Yaf9fPHMIdk.

## Other related files

- reference transcriptome from 10x Genomics: `/dcs04/lieber/lcolladotor/annotationFiles_LIBD001/10x/refdata-gex-GRCh38-2020-A/`
