---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r 'setup', include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
library('BiocStyle')
library('spatialLIBD')

## Bib setup
library('knitcitations')
## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')

# Note links won't show for now due to the following issue
# https://github.com/cboettig/knitcitations/issues/63
## Write bibliography information
bib <- c(
    R = citation(),
    AnnotationHub = citation('AnnotationHub'),
    BiocStyle = citation('BiocStyle'),
    cowplot = citation('cowplot'),
    DT = citation('DT'),
    ExperimentHub = citation('ExperimentHub'),
    fields = citation('fields'),
    ggplot2 = citation('ggplot2'),
    golem = citation('golem'),
    IRanges = citation('IRanges'),
    knitcitations = citation('knitcitations'),
    knitr = citation('knitr')[3],
    plotly = citation('plotly'),
    Polychrome = citation('Polychrome'),
    RColorBrewer = citation('RColorBrewer'),
    rmarkdown = citation('rmarkdown')[1],
    S4Vectors = citation('S4Vectors'),
    scater = citation('scater'),
    sessioninfo = citation('sessioninfo'),
    SingleCellExperiment = citation('SingleCellExperiment'),
    shiny = citation('shiny'),
    SummarizedExperiment = citation('SummarizedExperiment'),
    testthat = citation('testthat'),
    viridisLite = citation('viridisLite'),
    spatialLIBD = citation('spatialLIBD')[1],
    spatialLIBDpaper = citation('spatialLIBD')[2]
)
```

# Spatial DLPFC

This project has the code for the project using 10X Genomics Visium for generating spatial transcriptomics data from the dorsolateral prefrontal cortex (DLPFC) from human brain samples. This work is being done by the groups at the Lieber Institute for Brain Development led by Keri Martinowich, Kristen Maynard and Leonardo Collado-Torres as well as Stephanie Hicks from JHBSPH's Biostatistics Department.

##Study iesign

##Shiny website mirrors

## Installation instructions

Get the latest stable `R` release from
[CRAN](http://cran.r-project.org/). Then install `spatialLIBD` using
from [Bioconductor](http://bioconductor.org/) the following code:

```{r 'install', eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("spatialLIBD")
```

## Access the data

Through the `spatialLIBD` package you can access the processed data in
itâ€™s final R format.

### Processed data

Using `spatialLIBD` you can access the Human DLPFC spatial
transcriptomics data from the 10x Genomics Visium platform. For example,
this is the code you can use to access the spatially-resolved data. For more
details, check the help file for `fetch_data()`.

```{r 'access_data', message=FALSE, fig.height = 8, fig.width = 9, eval = FALSE}
## Load the package
library("spatialLIBD")
## Download the spot-level data
spe <- fetch_data(type = "spe_spatialDLPFC") ## TODO update spatialLIBD
#> Loading objects:
#>   spe
## This is a SpatialExperiment object
spe ## TODO: update the output below
#> class: SingleCellExperiment 
#> dim: 33538 47681 
#> metadata(1): image
#> assays(2): counts logcounts
#> rownames(33538): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
#>   ENSG00000268674
#> rowData names(9): source type ... gene_search is_top_hvg
#> colnames(47681): AAACAACGAATAGTTC-1 AAACAAGTATCTCCCA-1 ...
#>   TTGTTTCCATACAACT-1 TTGTTTGTGTAAATTC-1
#> colData names(73): barcode sample_name ... pseudobulk_UMAP_spatial
#>   markers_UMAP_spatial
#> reducedDimNames(6): PCA TSNE_perplexity50 ... TSNE_perplexity80
#>   UMAP_neighbors15
#> spikeNames(0):
#> altExpNames(0):
## Note the memory size
lobstr::object_size(spe) 
#> 2.08 GB ## TODO: update this
## Remake the logo image with histology information
spatialLIBD::vis_clus(
    spe = spe,
    clustervar = "bayesSpace_harmony_9",
    sampleid = "Br2720_Ant",
    colors = libd_layer_colors, ## TODO
    ... = " spatial DLPFC Human Brain k=9\nMade with github.com/LieberInstitute/spatialLIBD"
)
```

TODO: update this:

<img src="img/study_overview.png" width="600px" align="center" />

### Raw data

You can access all the raw data through [Globus](http://research.libd.org/globus/) (`jhpce#spatialDLPFC`).


# Internal

_Note: the idea is to use the same structure as in the [`locus-c`](https://github.com/lmweber/locus-c) project._

* JHPCE location: `/dcs04/lieber/lcolladotor/spatialDLPFC_LIBD4035/spatialDLPFC`.
* Slack channel: `libd_dlpfc_spatial`.

## Files

- `code`: R, python, and shell scripts for running various analyses.
  * `spot_deconvo`: cell-type deconvolution within Visium spots, enabled by tools like `tangram`, `cell2location`, and `cellpose`
  * `spython`: older legacy testing scripts mostly replaced by `spot_deconvo`
- `plots`: plots generated by RMarkdown or R analysis scripts in `.pdf` or `.png` format
- `processed-data`
  * `images_spatialLIBD`: images used for running `SpaceRanger`
  * `NextSeq`: `SpaceRanger` output files
  * `rdata`: R objects
- `raw-data`
  * `FASTQ`: FASTQ files from `NextSeq` runs.
  * `FASTQ_renamed`: renamed symbolic links to the original FASTQs, with consistent nomenclature
  * `Images`: raw images from the scanner in `.tif` format and around 3 GB per sample.
  * `images_raw_align_json`
  * `psychENCODE`: external data from PsychENCODE, originally retrieved from [here](https://www.synapse.org/#!Synapse:syn30108587.1/datasets/)
  * `sample_info`: spreadsheet with information about samples (sample ID, sample name, slide serial number, capture area ID)
  
This project is organized along the guidelines at https://lcolladotor.github.io/bioc_team_ds/organizing-your-work.html#.Yaf9fPHMIdk.

## Other related files

- reference transcriptome from 10x Genomics: `/dcs04/lieber/lcolladotor/annotationFiles_LIBD001/10x/refdata-gex-GRCh38-2020-A/`
