---
title: "LC analyses: features per spot (LC samples)"
author: "Lukas Weber"
date: "19 Nov 2020"
output:
  html_document:
    toc: true
    number_sections: true
    toc_depth: 3
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction

Analyses for LC project: number of features per spot.

One of the key characteristics of the LC Visium samples is that most spots contain very small numbers of features (UMI counts or expressed genes) per spot, and a small region of spots contains much higher numbers of features.

Our initial thoughts are that this could be either due to:

1. LC neurons are concentrated in the smaller region, so these spots capture orders of magnitude higher numbers of mRNAs than the "background" spots consisting largely of axon tracts / fiber bundles in white matter.

2. Experimental problems, e.g. with capturing mRNAs in the "background" spots.

This file contains analyses and plots to investigate these questions.


## Data

In this file, we are using the combined NextSeq and MiSeq reads for the LC samples (LC_1 and LC_2).


# Analysis

## Load data

Load data from Space Ranger outputs into R.

Notes about Space Ranger outputs:

- "raw" feature-barcode matrix contains all spots
- "filtered" feature-barcode matrix contains only spots identified as being under tissue
- more details: https://support.10xgenomics.com/spatial-gene-expression/software/pipelines/latest/output/matrices

Here we use the "filtered" feature-barcode matrix to create a `SingleCellExperiment`.

In the future, we will use `SpatialExperiment` instead (once the class structure is finalized): https://bioconductor.org/packages/SpatialExperiment

```{r, message=FALSE}
library(SingleCellExperiment)
library(Matrix)
library(rjson)
```

```{r}
sample_names <- c("DLPFC_3942_ant", "DLPFC_3942_mid","DLPFC_3942_post", "DLPFC_Br2743_ant", "DLPFC_Br2743_mid", "DLPFC_Br2743_post", "DLPFC_Br6423_ant", "DLPFC_Br6423_mid", "DLPFC_Br6423_post", "DLPFC_Br8492_ant", "DLPFC_Br8492_mid", "DLPFC_Br8492_post")
sce_list <- vector("list", length = length(sample_names))
names(sce_list) <- sample_names

for (i in seq_along(sample_names)) {
  
  # ---------
  # load data
  # ---------
  
  # select sample
  sample_name <- sample_names[i]
  
  # path to Space Ranger output files
  if (Sys.info()["sysname"] == "Linux") {
    # files on JHPCE cluster
    dir_outputs <- "/dcl02/lieber/ajaffe/SpatialTranscriptomics/LIBD/spatialDLPFC/outputs/NextSeq"
  } else if (Sys.info()["sysname"] == "Darwin") {
    # copy of files on Mac laptop
    dir_outputs <- "~/DLPFC"
  }
  
  # note: using "filtered" barcodes list containing only spots over tissue
  dir_matrix <- file.path(dir_outputs, sample_name, "outs", "raw_feature_bc_matrix")
  
  # barcodes
  file_barcodes <- file.path(dir_matrix, "barcodes.tsv.gz")
  df_barcodes <- read.csv(file_barcodes, sep = "\t", header = FALSE, 
                          col.names = c("barcode_id"))
  # features
  file_features <- file.path(dir_matrix, "features.tsv.gz")
  df_features <- read.csv(file_features, sep = "\t", header = FALSE, 
                          col.names = c("gene_id", "gene_name", "feature_type"))
  # counts
  file_counts <- file.path(dir_matrix, "matrix.mtx.gz")
  counts <- readMM(file = file_counts)
  
  # spatial coordinates
  dir_spatial <- file.path(dir_outputs, sample_name, "outs", "spatial")
  file_tisspos <- file.path(dir_spatial, "tissue_positions_list.csv")
  df_tisspos <- read.csv(file_tisspos, header = FALSE, 
                         col.names=c("barcode_id", "in_tissue", "array_row", "array_col", 
                                     "pxl_col_in_fullres", "pxl_row_in_fullres"))
  
  # check dimensions
  dim(df_barcodes)
  dim(df_features)
  dim(counts)
  # note df_tisspos contains all spots (not filtered) - need to match later
  dim(df_tisspos)
  
  # image paths
  imageFilePath <- file.path(dir_spatial, c("tissue_hires_image.png", "tissue_lores_image.png"))
  
  # spatial scale factors
  file_scale <- file.path(dir_spatial, "scalefactors_json.json")
  scalefactors <- fromJSON(file = file_scale)
  
  
  # ---------------------------
  # create SingleCellExperiment
  # ---------------------------
  
  # note: check and/or re-order rows to make sure barcode IDs match in df_barcodes and df_tisspos
  dim(df_barcodes)
  dim(df_tisspos)
  ord <- match(df_barcodes$barcode_id, df_tisspos$barcode_id)
  df_tisspos_ord <- df_tisspos[ord, ]
  dim(df_tisspos_ord)
  stopifnot(nrow(df_barcodes) == nrow(df_tisspos_ord))
  stopifnot(all(df_barcodes$barcode_id == df_tisspos_ord$barcode_id))
  
  head(df_barcodes)
  head(df_tisspos_ord)
  
  col_data <- cbind(df_barcodes, df_tisspos_ord[, -1])
  head(col_data)
  
  sce <- SingleCellExperiment(
    rowData = df_features, 
    colData = col_data, 
    assays = c(counts = counts), 
    metadata = list(scalefactors = scalefactors)
  )
  
  sce
  
  # store object
  sce_list[[i]] <- sce
}

sce_list

save(sce_list, file = "~/DLPFC/sce_list.RDS")
```
#> sce_list
$DLPFC_3942_ant
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_3942_mid
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_3942_post
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br2743_ant
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br2743_mid
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br2743_post
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br6423_ant
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br6423_mid
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br6423_post
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br8492_ant
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br8492_mid
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

$DLPFC_Br8492_post
class: SingleCellExperiment
dim: 36601 4992
metadata(1): scalefactors
assays(1): counts
rownames: NULL
rowData names(3): gene_id gene_name feature_type
colnames: NULL
colData names(6): barcode_id in_tissue ... pxl_col_in_fullres
  pxl_row_in_fullres
reducedDimNames(0):
altExpNames(0):

## Calculate scater QC metrics


```{r, message=FALSE}
library(here)
library(scater)
library(ggplot2)
```

# ```{r message=FALSE}
# for (i in seq_along(sample_names)) {
#   
#   # select sample
#   sce <- sce_list[[i]]
#   
#   # identify mitochondrial genes
#   is_mito <- grepl("(^MT-)|(^mt-)", rowData(sce)$gene_name)
#   table(is_mito)
#   rowData(sce)$gene_name[is_mito]
#   
#   # calculate QC metrics using scater package
#   sce <- addPerCellQC(sce, subsets = list(mito = is_mito))
#   
#   colData(sce)
#   
#   # store
#   sce_list[[i]] <- sce
# }
# ```


## Percent mitochondrial counts

# ```{r, message=FALSE, fig.width=4.75, fig.height=3.5}
# # plot proportion of mitochondrial counts by spot
# for (i in seq_along(sample_names)) {
#   
#   # select sample
#   sce <- sce_list[[i]]
#   
#   # plot histogram
#   hist(colData(sce)$subsets_mito_percent, 
#        main = paste0("Percent mitochondrial counts by spot\nsample ", names(sce_list[i])), 
#        xlab = "percent mitochondrial counts")
# }
# ```


## Top expressed genes

# ```{r, message=FALSE}
# # identify top expressed genes for each sample
# 
# for (i in seq_along(sample_names)) {
#   
#   # select sample
#   sce <- sce_list[[i]]
#   
#   # identify top expressed genes
#   sum_counts_by_gene <- rowSums(counts(sce))
#   print(names(sce_list)[i])
#   print(rowData(sce)[rank(sum_counts_by_gene, ties.method = "first") <= 10, ])
# }
# ```


## Number of features per spot

Spot-level summaries from Space Ranger output file `web_summary.html` for each sample.

Note that we are using combined reads from the NextSeq and MiSeq runs, where 97-99% of total reads are from NextSeq.

Additional plots comparing against the additional DLPFC sample (where only MiSeq reads are available) are included in the other analysis file `features_per_spot_MiSeq.Rmd`.

LC_1:

- Number of Spots Under Tissue: 2,979
- Fraction Reads in Spots Under Tissue: 91.2%
- Mean Reads per Spot: 111,630
- Median Genes per Spot: 235
- Total Genes Detected: 19,352
- Median UMI Counts per Spot: 330

- Total reads from NextSeq: 327,107,961 (98.4% of total)
- Total reads from MiSeq: 5,437,332 (1.6% of total)
- Total reads: 332,545,293

LC_2:

- Number of Spots Under Tissue: 3,056
- Fraction Reads in Spots Under Tissue: 82.4%
- Mean Reads per Spot: 86,002
- Median Genes per Spot: 83
- Total Genes Detected: 17,223
- Median UMI Counts per Spot: 118

- Total reads from NextSeq: 255,840,579 (97.3% of total)
- Total reads from MiSeq: 6,982,110 (2.7% of total)
- Total reads: 262,822,689


## Number of UMI counts per spot

Thresholds for the regions with high UMI counts were selected by checking plots with various thresholds, to find values that most clearly distinguish between the two regions.

```{r, message=FALSE, fig.width=4.75, fig.height=3.5}
# number of UMI counts per spot in different regions of each sample

dir_plot <- here("plots", "features_per_spot")
if (!dir.exists(dir_plot)) dir.create(dir_plot, recursive = TRUE)

thresh_counts_high <- c(500, 250)

for (i in seq_along(sample_names)) {
  
  # select sample
  sce <- sce_list[[i]]
  
  # identify spots with high/low counts
  colData(sce)$is_high_counts <- colData(sce)$sum > thresh_counts_high[i]
  # store object
  sce_list[[i]] <- sce
  
  
  # plot UMI counts per spot
  
  p <- ggplot(as.data.frame(colData(sce)), 
              aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = sum)) + 
    geom_point(size = 0.5) + 
    coord_fixed() + 
    scale_y_reverse() + 
    scale_color_gradient(low = "gray95", high = "navy") + 
    ggtitle(paste0("UMI counts per spot: ", sample_names[i])) + 
    theme_bw() + 
    theme(panel.grid = element_blank(), 
          axis.title = element_blank(), 
          axis.text = element_blank(), 
          axis.ticks = element_blank())
  
  print(p)
  
  fn <- file.path(dir_plot, paste0("xyplot_UMI_counts_per_spot_", sample_names[i], ".png"))
  ggsave(fn, width = 4.75, height = 3.5)
  
  
  # plot spots with high UMI counts
  
  p <- ggplot(as.data.frame(colData(sce)), 
              aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = is_high_counts)) + 
    geom_point(size = 0.5) + 
    coord_fixed() + 
    scale_y_reverse() + 
    scale_color_manual(values = c("gray95", "red")) + 
    ggtitle(paste0("Spots with UMI counts > ", thresh_counts_high[i], ": ", sample_names[i])) + 
    theme_bw() + 
    theme(panel.grid = element_blank(), 
          axis.title = element_blank(), 
          axis.text = element_blank(), 
          axis.ticks = element_blank())
  
  print(p)
  
  fn <- file.path(dir_plot, paste0("xyplot_high_UMI_counts_", sample_names[i], ".png"))
  ggsave(fn, width = 4.75, height = 3.5)
}
```


## Number of detected genes per spot

Thresholds for the regions with high numbers of detected genes were selected by checking plots with various thresholds, to find values that most clearly distinguish between the two regions.

```{r, message=FALSE, fig.width=4.75, fig.height=3.5}
# number of detected genes per spot in different regions of each sample

thresh_genes_high <- c(2000, 1000)

for (i in seq_along(sample_names)) {
  
  # select sample
  sce <- sce_list[[i]]
  
  # identify spots with high/low number of detected genes
  colData(sce)$is_high_n_genes <- colData(sce)$detected > thresh_genes_high[i]
  # store object
  sce_list[[i]] <- sce
  
  
  # plot number of genes detected per spot
  
  p <- ggplot(as.data.frame(colData(sce)), 
              aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = detected)) + 
    geom_point(size = 0.5) + 
    coord_fixed() + 
    scale_y_reverse() + 
    scale_color_gradient(low = "gray95", high = "darkgreen") + 
    ggtitle(paste0("Detected genes per spot: ", sample_names[i])) + 
    theme_bw() + 
    theme(panel.grid = element_blank(), 
          axis.title = element_blank(), 
          axis.text = element_blank(), 
          axis.ticks = element_blank())
  
  print(p)
  
  fn <- file.path(dir_plot, paste0("xyplot_n_genes_per_spot_", sample_names[i], ".png"))
  ggsave(fn, width = 4.75, height = 3.5)
  
  
  # plot spots with high number of detected genes
  
  p <- ggplot(as.data.frame(colData(sce)), 
              aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = is_high_n_genes)) + 
    geom_point(size = 0.5) + 
    coord_fixed() + 
    scale_y_reverse() + 
    scale_color_manual(values = c("gray95", "red")) + 
    ggtitle(paste0("Spots with no. detected genes > ", thresh_genes_high[i], ": ", sample_names[i])) + 
    theme_bw() + 
    theme(panel.grid = element_blank(), 
          axis.title = element_blank(), 
          axis.text = element_blank(), 
          axis.ticks = element_blank())
  
  print(p)
  
  fn <- file.path(dir_plot, paste0("xyplot_high_n_genes_", sample_names[i], ".png"))
  ggsave(fn, width = 4.75, height = 3.5)
}
```


## Marker genes

Investigate expression of known marker genes for norepinephrine cells (LC neurons): tyrosine hydroxylase (gene name "TH") and norepinephrine transporter (NET; gene name "SLC6A2").

Note that these plots are using raw counts. Log-transformed normalized counts in spots meeting QC metrics will be calculated later for downstream analyses, after segmenting out the region containing high counts.

```{r, message=FALSE, fig.width=4.75, fig.height=3.5}
# expression of known marker genes using raw UMI counts

marker_gene_names <- c("MAP2", "TH", "SLC6A2")
marker_genes_thresh <- 2

for (i in seq_along(sample_names)) {
  
  # select sample
  sce <- sce_list[[i]]
  
  for (j in seq_along(marker_gene_names)) {
    
    # identify marker gene
    ix_marker <- which(rowData(sce)$gene_name == marker_gene_names[j])
    colData(sce)$counts_marker <- counts(sce)[ix_marker, ]
    colData(sce)$is_counts_marker <- counts(sce)[ix_marker, ] > marker_genes_thresh
    
    
    # plot UMI counts for marker gene
    
    p <- ggplot(as.data.frame(colData(sce)), 
                aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = counts_marker)) + 
      geom_point(size = 0.5) + 
      coord_fixed() + 
      scale_y_reverse() + 
      scale_color_gradient(low = "gray95", high = "purple4") + 
      ggtitle(paste0("UMI counts for gene ", marker_gene_names[j], ": ", sample_names[i])) + 
      theme_bw() + 
      theme(panel.grid = element_blank(), 
            axis.title = element_blank(), 
            axis.text = element_blank(), 
            axis.ticks = element_blank())
    
    print(p)
    
    fn <- file.path(dir_plot, paste0("xyplot_counts_", marker_gene_names[j], "_", sample_names[i], ".png"))
    ggsave(fn, width = 4.75, height = 3.5)
    
    
    # plot spots with UMI counts above threshold for marker gene
    
    p <- ggplot(as.data.frame(colData(sce)), 
                aes(x = pxl_row_in_fullres, y = pxl_col_in_fullres, color = is_counts_marker)) + 
      geom_point(size = 0.5) + 
      coord_fixed() + 
      scale_y_reverse() + 
      scale_color_manual(values = c("gray95", "purple4")) + 
      ggtitle(paste0("Spots with UMI counts > ", marker_genes_thresh, " for gene ", marker_gene_names[j], ": ", sample_names[i])) + 
      theme_bw() + 
      theme(panel.grid = element_blank(), 
            axis.title = element_blank(), 
            axis.text = element_blank(), 
            axis.ticks = element_blank())
    
    print(p)
    
    fn <- file.path(dir_plot, paste0("xyplot_counts_thresh_", marker_gene_names[j], "_", sample_names[i], ".png"))
    ggsave(fn, width = 4.75, height = 3.5)
  }
}
```


# Session info

```{r, message=FALSE}
sessionInfo()
```
