# tangram-libd

This repository is our current and in-progress implementation of [the Broad Institute's Tangram package](https://github.com/broadinstitute/Tangram) for spatial transcriptomic imputation by Biancalani* T., Scalia* G. et al. You can find their publication on this package here:

Biancalani* T., Scalia* G. et al. - _Deep learning and alignment of spatially-resolved whole transcriptomes of single cells in the mouse brain with Tangram_  [biorXiv 10.1101/2020.08.29.272831](https://www.biorxiv.org/content/10.1101/2020.08.29.272831v3) (2020)

## Requirements
* R/4.0.4
* Python/3.7.3
* SGE 8.1.9
* [See Tangram's `yaml` for Tangram-specific dependencies](https://github.com/broadinstitute/Tangram/blob/master/environment.yml)

## Directory Structure
### ./archive
Contains scripts pertaining to previous versions of the `tangram-libd` pipeline. These scripts are purely for reference and are no longer of any functional importance.

### ./data
Contains data generated by the pipeline itself to be used further downstream. These files may also be of interest in themselves:
* `brain_samples.txt` is a vector of brain sample IDs that is used to restrict the spatial dataset in `sce_to_anndata.R` to only those which are going to be tested in the Tangram run.
* `marker_stats.csv` is a one-column list of the most highly-expressed cell-type-specific genes, totaling to 1,000.

###  ./scripts
Tangram pipeline lives here. TBD

### ./out
This directory contains the outputs of the Tangram pipeline, where plots and tables will be located.
